src/segment-loader.js

```diff
@@ -1178,6 +1221,187 @@ export default class SegmentLoader extends videojs.EventTarget {
+   * @param {Object} event.target The SourceBuffer that caused the error
+   */
+  handleBufferMaxed_(event) {
+    let bytesTried = event.segment.byteLength;
+    let segmentBytes = event.segment;
+    let sourceBuffer = event.target;
+    let safeRemovePoint = event.safeRemovePoint;
+    let pendingAppend = true;
+    let callback = event.done;
+
+    // if we fail on the first request, it's probably a glitch, and
+    // we should just slice the segment up into small bits and append
+    if (this.mediaRequests === 1) {
+      let slices = [];
+      let start = 0;
+      let end = 0;
+      let sliceAppend = () => {
+        if (slices.length) {
+          sourceBuffer.appendBuffer(slices.shift());
+        } else {
+          sourceBuffer.removeEventListener('updateend', sliceAppend);
+          callback();
+        }
+      };
+
+      while (start < bytesTried) {
+        end += 5000000;
+        if (end > bytesTried) {
+          end = bytesTried;
+        }
+        slices.push(segmentBytes.slice(start, end));
+        start += 5000000;
+      }
+      sourceBuffer.addEventListener('updateend', sliceAppend);
+      sourceBuffer.appendBuffer(slices.shift());
+      return;
+    }
+
+    let bytesTrimmable;
+    let removeToTime;
+    let timeupdateHandler;
+    let tryAppend = () => {
+      if (!pendingAppend) {
+        return;
+      }
+      this.hls_.tech_.off('timeupdate', timeupdateHandler);
+      this.logger_('Appended deferred segment with ' +
+        (this.buffered_().end(0) - this.currentTime_()) + ' to spare...');
+      // we don't want this to be called again
+      sourceBuffer.removeEventListener('updateend', tryAppend);
+      try {
+        sourceBuffer.appendBuffer(segmentBytes);
+        callback();
+        pendingAppend = false;
+      } catch (error) {
+        videojs.error('QuoteExceededError');
+      }
+    };
+
+    sourceBuffer.addEventListener('updateend', tryAppend);
+
+    if (safeRemovePoint) {
+      removeToTime = safeRemovePoint;
+    } else {
+      removeToTime = safeBackBufferTrimTime(this.seekable_(),
+                                              this.currentTime_(),
+                                              this.playlist_.targetDuration || 10,
+                                              this.backBufferLength_());
+    }
+
+    bytesTrimmable = this.minBufferedBytes(0, removeToTime);
+
+    if (bytesTrimmable >= bytesTried) {
+      this.removeWholeSegments(removeToTime);
+    } else {
+      let waitUntil = this.currentTime_() + this.playlist_.targetDuration;
+      let buffered = this.buffered_();
+      let dangerZone = buffered.end(buffered.length - 1) - 2;
+
+      timeupdateHandler = () => {
+        let currentTime = this.currentTime_();
+
+        if (currentTime < waitUntil) {
+          return;
+        } else if (currentTime > dangerZone) {
+          this.removeWholeSegments(currentTime - this.playlist_.targetDuration);
```

Comment:
The idea of the `dangerZone` is that it's one last hail mary to try to unblock the append. Unfortunately, it's pretty much worthless without GOP slicing, since if you're operating that near the threshold of getting stuck, your back buffer is probably already at its lowest possible value -- a target duration.

It removes any whole segments that are within the time range [0, currentTime - targetDuration].